mux/src/lib.rs

```diff
@@ -1439,3 +1551,106 @@ impl wezterm_term::DownloadHandler for MuxDownloader {
         }
     }
 }
+
+fn cmd_cwd_from_pane(
+    domain: &dyn Domain,
+    p: &SavedPaneState,
+) -> (CommandBuilder, Option<String>, Option<String>) {
+    let cwd = if domain.is::<LocalDomain>() {
+        p.cwd.as_ref().and_then(|u| {
+            percent_decode_str(u.url.path())
+                .decode_utf8()
+                .ok()
+                .map(|path| path.into_owned())
+        })
+    } else {
+        None
+    };
+    let mut cmd = CommandBuilder::new(&p.root_argv[0]);
+    cmd.args(&p.root_argv[1..]);
+    let prog = p
+        .prog_argv
+        .as_ref()
+        .and_then(|p| shlex::try_join(p.iter().map(|s| s.as_str())).ok());
+    (cmd, cwd, prog)
+}
+
+fn leftmost_pane(t: &SavedTree) -> &SavedPaneState {
+    match t {
+        SavedTree::Pane(p) => p.as_ref().unwrap(),
+        SavedTree::Split(s) => leftmost_pane(&s.left),
+    }
+}
+
+fn split_panes<'m>(
+    domain: &'m dyn Domain,
+    split: SavedTree,
+    tab: TabId,
+    pane: PaneId,
+    id_map: &'m mut HashMap<usize, usize>,
+) -> Pin<Box<dyn Future<Output = ()> + 'm>> {
+    Box::pin(async move {
+        if let SavedTree::Split(s) = split {
+            let leftmost = leftmost_pane(&s.right);
+            let (cmd, cwd, prog) = cmd_cwd_from_pane(domain, &leftmost);
+            let source = SplitSource::Spawn {
+                command: Some(cmd),
+                command_dir: cwd,
+            };
+            let req = SplitRequest {
+                direction: s.dir,
+                target_is_second: true,
+                top_level: false,
+                size: SplitSize::Cells(s.pos),
+            };
+            let new = domain.split_pane(source, tab, pane, req).await.unwrap();
+            if let Some(prog) = prog {
+                new.writer().write_all(prog.as_bytes()).unwrap();
+                new.writer().write_all(b"\r").unwrap();
+            }
+            assert_eq!(id_map.insert(leftmost.id, new.pane_id()), None);
+            split_panes(domain, *s.left, tab, pane, id_map).await;
+            split_panes(domain, *s.right, tab, new.pane_id(), id_map).await;
+        }
+    })
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+pub struct SavedPaneState {
+    id: usize,
+    cwd: Option<SerdeUrl>,
+    root_argv: Vec<String>,
```

Comment:
the domain and other important info are missing from this, but I think the approach here needs to be different; I'll cover this more in the main overall review comment